lib/evm/instruction/impl.ex

```diff
@@ -50,240 +51,208 @@ defmodule EVM.Instruction.Impl do
-  def add([s0, s1], %{stack: stack}) do
-    stack |> push(s0 + s1)
+  def add([s0, s1], _) do
+    s0 + s1
+      |> wrap_int
+      |> encode_signed
   end
 
   @doc """
   Multiplication operation.
 
   ## Examples
 
-      iex> EVM.Instruction.Impl.mul([5, 2], %{stack: []})
-      %{stack: [10]}
-
-      iex> EVM.Instruction.Impl.mul([-1, 5], %{stack: []})
-      %{stack: [-5]}
+      iex> EVM.Instruction.Impl.mul([5, 2], %{})
+      10
 
-      iex> EVM.Instruction.Impl.mul([EVM.max_int() + 1, 5], %{stack: []})
-      %{stack: [5]}
+      iex> EVM.Instruction.Impl.mul([-1, 5], %{})
+      EVM.Instruction.Impl.encode_signed(-5)
   """
   @spec mul(stack_args, vm_map) :: op_result
-  def mul([s0, s1], %{stack: stack}) do
-    stack |> push(s0 * s1)
+  def mul([s0, s1], _) do
+    s0 * s1
+      |> wrap_int
+      |> encode_signed
   end
 
   @doc """
   Subtraction operation.
 
   ## Examples
 
-      iex> EVM.Instruction.Impl.sub([5, 2], %{stack: []})
-      %{stack: [3]}
+      iex> EVM.Instruction.Impl.sub([5, 2], %{})
+      3
 
-      iex> EVM.Instruction.Impl.sub([-1, 5], %{stack: []})
-      %{stack: [-6]}
-
-      iex> EVM.Instruction.Impl.sub([EVM.max_int() + 6, 5], %{stack: []})
-      %{stack: [1]}
+      iex> EVM.Instruction.Impl.sub([-1, 5], %{})
+      EVM.Instruction.Impl.encode_signed(-6)
   """
   @spec sub(stack_args, vm_map) :: op_result
-  def sub([s0, s1], %{stack: stack}) do
-    stack |> push(s0 - s1)
+  def sub([s0, s1], _) do
+    s0 - s1
+      |> wrap_int
+      |> encode_signed
   end
 
   @doc """
   Integer division operation.
 
   ## Examples
 
-      iex> EVM.Instruction.Impl.div([5, 2], %{stack: []})
-      %{stack: [2]}
-
-      iex> EVM.Instruction.Impl.div([5, -2], %{stack: []})
-      %{stack: [-3]}
+      iex> EVM.Instruction.Impl.div([5, 2], %{})
+      2
 
-      iex> EVM.Instruction.Impl.div([10, 2], %{stack: []})
-      %{stack: [5]}
+      iex> EVM.Instruction.Impl.div([10, 2], %{})
+      5
 
-      iex> EVM.Instruction.Impl.div([10, 0], %{stack: []})
-      %{stack: [0]}
-
-      iex> EVM.Instruction.Impl.div([EVM.max_int() + 6, 1], %{stack: []})
-      %{stack: [6]}
+      iex> EVM.Instruction.Impl.div([10, 0], %{})
+      0
   """
-  def div([_s0, 0], %{stack: stack}), do: push(stack, 0)
-  def div([s0, s1], %{stack: stack}) do
-    stack |> push(Integer.floor_div(s0, s1))
+  def div([_s0, 0], _), do: 0
+  def div([s0, s1], _) do
+    Integer.floor_div(s0, s1)
+      |> wrap_int
+      |> encode_signed
   end
 
   @doc """
   Signed integer division operation (truncated).
-
-  TODO: Implement opcode
-
-  ## Examples
-
-      iex> EVM.Instruction.Impl.sdiv([], %{stack: []})
-      :unimplemented
   """
   @spec sdiv(stack_args, vm_map) :: op_result
-  def sdiv(_args, %{stack: _stack}) do
-    :unimplemented
+  def sdiv([s0, s1], _) do
+    case decode_signed(s1) do
+      0 ->
+          0
+      1 ->
+        s0
+      -1 ->
+        -decode_signed(s0)
```

Comment:
maybe `0 - decode_signed(s0)` or just something to make this more clear it's an additive inverse